### ISSUE TITLE
Add modular active learning

### ISSUE THREAD

**Issue by cisprague** [alan-turing-institute/autoemulate#301]
**Main idea**
1. Implement an active learning loop where input points are selected to query the simulation model, retrieve output points, and iteratively improve the emulator model(s). A sampling budget should also be taken into account.
2. Modularity: The setup should be agnostic to the specific active learning approach, ensuring compatibility with history matching (#294) and Bayesian methods, while also supporting deterministic emulator models (e.g., using dropout for uncertainty estimation, or heuristic policies).
3. Since each learner (or "agent") could have multiple emulators, we need to ensure that policy is designed to handle multi-emulator scenarios. It would also be nice to leave open the possiblity of the policy being learnable.

**Possible API (psuedocode)**
- Here's a proposed API for an active learning loop (with 1 emulator) that aims to support multiple methods. If any existing components should be integrated, let me know. Would love to hear thoughts.
- Multiple emulators could possibly be accomodated by [EnsembleModel](https://github.com/pytorch/botorch/blob/052d0d8046c101504a01cac28f3f56e8fa0c6fda/botorch/models/ensemble.py#L24), and deterministic models (e.g. dropouot) by [EnsemblePosterior](https://github.com/pytorch/botorch/blob/052d0d8046c101504a01cac28f3f56e8fa0c6fda/botorch/posteriors/ensemble.py#L18), also see [dropout NN posterior](https://github.com/pytorch/botorch/discussions/2127#discussion-5880474).

```python
class Learner:

    def __init__(self, emulator, simulator, input_space, max_obs):
        self.emulator = emulator # Surrogate model
        self.simulator = simulator # Callable wrapper around user's sim
        self.obs = [] # Buffer of input-output pairs
        self.max_obs = max_obs # Max number of observations to store

    def update(self, new_obs):
        # Could have different update logic here
        if len(self.obs) < self.max_obs:
            self.obs.append(new_obs)
        else:
            idx = random.randint(0, self.max_obs - 1)
            self.obs[idx] = new_obs

    def policy(self):
        # Method to select new query points, which will be
        # different for different strategies.
        # Should be adapted to multiple emulators
        raise NotImplementedError # or ABC

    def iterate(self, simulation_func):
        # One iteration of active learning
        self.emulator.fit()
        new_point = self.policy()
        new_output = self.simulator(new_point)
        self.update((new_point, new_output))

class Bayesian_Optimizer(Learner):

    def __init__(self, emulator, simulator, input_space, max_obs, acquisition_func):
        super().__init__(emulator, simulator, input_space, max_obs)
        self.acquisition_fn = acquisition_func  # e.g., Expected Improvement

    def policy(self):
        return self.acquisition_fn(self.emulator)

class History_Matcher(Learner):

    def __init__(self, emulator, simulator, input_space, max_obs, history_matcher):
        super().__init__(emulator, simulator, input_space, max_obs)
        self.history_matcher = history_matcher

    def not_ruled_out_yet(self, input_space):
        # Rejection sampling of NROY points
        candidates = latin_hypercube(input_space)
        mask = nroy_func(candidates)  # nroy_func applies the history matching filter.
        return candidates[mask]

    def policy(self):
        # Randomly sample from candidates that have not been ruled out.
        plausible = self.not_ruled_out_yet(self.input_space)
        return random.choice(plausible)
```

**Comment by mastoffel** [comment]
@all-contributors please add @cisprague for code, design, ideas

**Comment by cisprague** [comment]
@radka-j thanks for the questions! 

1. Re `iterate()`: Yea, the order is flexible (but I agree that your ordering makes the most sense). I originally wrote it that way since we might want to fit the model initially if we start with some pre-existing data -- but that can be handled in `__init__`.

Re `EnsembleX()`:
1. I think we could have both *single* and *ensemble* emulators inherit from a common `Emulator` base class. Then `self.emulator.predict()` can handle returning values that align with `self.policy()`. E.g., if our policy expects variance estimates, we should ensure that `predict()` provides them.
2. I'm not sure if there's a scenario where the policy would need access to each individual emulator in an ensemble -- maybe in a heterogeneous ensemble with both deterministic (e.g. MLP) and stochastic (e.g. GP) models? But I haven't seen this in the literature yet.

Re learning a policy:
1. Yes, it seems to have been done for at least MetaBO for transfer learning (e.g. https://arxiv.org/abs/1904.02642), where they first train the policy on a cheaper simulation and then use it in the real system. Another example is https://arxiv.org/abs/2305.15930.
2. I haven't seen a specific API for this setup, but Stable Baselines `learn` is a good reference (https://stable-baselines.readthedocs.io/en/master/modules/base.html).

**Comment by cisprague** [comment]
The API in #358 has been updated to:

```python
@dataclass(kw_only=True)
class Active(Learner):
    simulator: Simulator
    emulator: Emulator
    X_train: torch.Tensor
    Y_train: torch.Tensor

    def fit(self, *args):
        # Query simulator and fit emulator
        X, Y_pred, Sigma, extra = self.query(*args)
        if X is not None:
            Y_true = self.simulator.sample(X)
            self.X_train = torch.cat([self.X_train, X])
            self.Y_train = torch.cat([self.Y_train, Y_true])
            self.emulator.fit(self.X_train, self.Y_train)
            ...
```

It would be nice to have a way to run multiple active learners in parallel for experiments. Perhaps something like this could work:

```python
class Ensemble(Active):
    learners: List[Active]
    simulator: Simulator
    X_train: torch.Tensor
    Y_train: torch.Tensor

    def fit(self, *args):
        for learner in self.learners:
            learner.fit(*args)

# Simulator and starting data shouldn't be repeated, but emulator should.
ensemble = Ensemble(
    learners=[
        Active(simulator=None, emulator=Emulator(), X_train=None, Y_train=None),
        Active(simulator=None, emulator=Emulator(), X_train=None, Y_train=None)
    ],
    simulator=Simulator(...), X_train=X_train, Y_train=Y_train
)
```
